labrad/backend.py

```diff
@@ -65,15 +68,31 @@ def sendMessage(self, target, records, *args, **kw):
 
     class TwistedConnection(BaseConnection):
         def _connect(self, password, _timeout, tls_mode):
+            @defer.inlineCallbacks
+            def _connect_deferred():
+                cxn = yield getConnection(self.host, self.port, self.name,
+                                          password, tls_mode=tls_mode)
+                self._connected.set()
+
+                @defer.inlineCallbacks
+                def await_disconnect():
+                    try:
+                        yield cxn.onDisconnect()
+                    except Exception:
+                        pass
+                    self._connected.clear()
+                await_disconnect()
```

Comment:
I don't want to yield here because that would cause this to block until the connection is dropped, so we'd never actually return in the thread where `connect` was called. I will make a comment about this in the code; also, the name `await_disconnect` is probably misleading in this regard.